u_parrots.go

```diff
@@ -619,38 +633,86 @@ func (uconn *UConn) generateRandomizedSpec(WithALPN bool) (ClientHelloSpec, erro
 		}
 		p.Extensions = append(p.Extensions, &ks, &pskExchangeModes, &supportedVersionsExt)
 	}
-	err = shuffleTLSExtensions(p.Extensions)
-	if err != nil {
-		return p, err
-	}
+	r.shuffleTLSExtensions(p.Extensions)
 	err = uconn.SetTLSVers(p.TLSVersMin, p.TLSVersMax)
 	if err != nil {
 		return p, err
 	}
 
-	return p, nil
+	return p, r.savedErr
+}
+
+// utlsRandomizer never returns errors, but stores first one in savedErr
+type utlsRandomizer struct {
+	r        io.Reader
+	savedErr error
+}
+
+func makeUtlsRandomizer(seed uint64) *utlsRandomizer {
+	seedBytes := make([]byte, 8)
+	binary.BigEndian.PutUint64(seedBytes, seed)
+	tdHkdf := hkdf.New(sha256.New, seedBytes, []byte("uTLS salty salt"), nil)
+	return &utlsRandomizer{r: tdHkdf}
+}
+
+// randUInt64 is the only function that touches source of entropy,
+// and other primitives are built using randUInt64
+func (r *utlsRandomizer) randUInt64() uint64 {
+	uintBytes := make([]byte, 8)
+	_, err := r.r.Read(uintBytes)
+	if r.savedErr == nil && err != nil {
+		r.savedErr = err
+	}
+	return binary.BigEndian.Uint64(uintBytes)
+}
+
+// returns int64 in [0, max)
+func (r *utlsRandomizer) randInt64n(max int) int64 {
+	if max == 0 {
+		return 0
+	}
+	randUint := r.randUInt64() % uint64(max)
+	if randUint > math.MaxInt64 {
+		randUint >>= 1
+	}
+	return int64(randUint)
+}
+
+// returns int in [0, max) up to max int32
+func (r *utlsRandomizer) randIntn(max int) int {
+	if max > math.MaxInt32 && strconv.IntSize < 64 {
+		// in practice, we never use randomizer with such big values for max
+		// if we ever do, it may panic on 32 bit architecture, so we have to check for that
+		r.savedErr = fmt.Errorf("utlsRandomizer does not support randIntn(%v): max is too big for a %v-bit architecture",
+			max, strconv.IntSize)
+		return 0
+	}
+	return int(r.randInt64n(max))
 }
 
-func tossBiasedCoin(probability float32) bool {
+func (r *utlsRandomizer) tossBiasedCoin(probability float32) bool {
 	// probability is expected to be in [0,1]
-	// this function never returns errors for ease of use
 	const precision = 0xffff
 	threshold := float32(precision) * probability
-	value, err := getRandInt(precision)
-	if err != nil {
-		// I doubt that this code will ever actually be used, as other functions are expected to complain
-		// about used source of entropy. Nonetheless, this is more than enough for given purpose
-		return ((time.Now().Unix() & 1) == 0)
-	}
-
+	value := r.randIntn(precision)
 	if float32(value) <= threshold {
 		return true
 	} else {
 		return false
 	}
 }
 
-func removeRandomCiphers(s []uint16, maxRemovalProbability float32) []uint16 {
+func (r *utlsRandomizer) getRandPerm(n int) []int {
```

Comment:
Use the runtime library's implementation.

Output is currently biased due to `r.randIntn` being broken.